src/jit/lir.cpp

```diff
@@ -578,6 +759,62 @@ bool LIR::Range::TryGetUse(GenTree* node, Use* use)
     return false;
 }
 
+//------------------------------------------------------------------------
+// LIR::Range::GetTreeRange: Computes the subrange that includes all nodes
+//                           in the dataflow tree rooted at a particular
+//                           node.
+//
+// Arguments:
+//    root     - The root of the dataflow tree.
+//    isClosed - An output parameter that is set to true if the returned
+//               range contains only nodes in the dataflow tree and false
+//               otherwise.
+//
+// Returns:
+//    The computed subrange.
+LIR::Range LIR::Range::GetTreeRange(GenTree* root, bool* isClosed) const
+{
+    assert(root != nullptr);
+    assert(isClosed != nullptr);
+
+    // Mark the root of the tree
+    unsigned markCount = 1;
+    root->gtLIRFlags |= LIR::Flags::Mark;
+
+    bool sawUnmarkedNode = false;
+
+    GenTree* firstNode;
+    for (firstNode = root; markCount > 0; firstNode = firstNode->gtPrev)
+    {
+        assert(firstNode != nullptr);
+
+        if ((firstNode->gtLIRFlags & LIR::Flags::Mark) != 0)
+        {
+            // Mark the node's operands
+            //
+            // TODO(pdg): replace with operand iterator
+            unsigned operandCount = firstNode->NumOperands();
+            for (unsigned i = 0; i < operandCount; i++)
+            {
+                GenTree* operand = *firstNode->GetOperand(i);
+                operand->gtLIRFlags |= LIR::Flags::Mark;
+                markCount++;
+            }
+
+            // Unmark the the node and udpate `firstNode`
+            firstNode->gtLIRFlags &= ~LIR::Flags::Mark;
+            markCount--;
+        }
+        else
+        {
+            sawUnmarkedNode = true;
+        }
+    }
+
+    *isClosed = !sawUnmarkedNode;
+    return LIR::AsRange(firstNode, root);
+}
```

Comment:
The algorithm depends on two interesting properties of LIR: (1) defs appear before uses, and (2) all defs exist before gtPrev goes null (meaning, essentially, all in the same block). Maybe useful to mention these things. How often does CheckLIR() get invoked? Maybe it should be invoked here?